genesis/engine/couplers/sap_coupler.py

```diff
@@ -1489,6 +1543,229 @@ def active_solvers(self):
+        )
+
+    @ti.func
+    def compute_constraint_regularization(self, sap_info, i_c, w_rms, time_step):
+        beta_factor = self.beta**2 / (4.0 * ti.math.pi**2)
+        dt2 = time_step**2
+        k = sap_info[i_c].k
+        R = max(beta_factor * w_rms, 1.0 / (dt2 * k))
+        sap_info[i_c].R = R
+        sap_info[i_c].R_inv = 1.0 / R
+
+    @ti.func
+    def compute_constraint_gamma_G(self, sap_info, i_c, vc):
+        y = (sap_info[i_c].v_hat - vc) * sap_info[i_c].R_inv
+        sap_info[i_c].gamma = y
+        sap_info[i_c].G = sap_info[i_c].R_inv
+
+    @ti.func
+    def compute_energy(self, energy: ti.template()):
+        constraints = ti.static(self.constraints)
+        sap_info = ti.static(constraints.sap_info)
+        for i_c in range(self.n_constraints[None]):
+            i_b = constraints[i_c].batch_idx
+            if self.coupler.batch_linesearch_active[i_b]:
+                vc = self.compute_vc(i_c)
+                self.compute_constraint_energy(sap_info, i_c, vc)
+                energy[i_b] += sap_info[i_c].energy
+
+    @ti.func
+    def compute_constraint_energy(self, sap_info, i_c, vc):
+        y = (sap_info[i_c].v_hat - vc) * sap_info[i_c].R_inv
+        sap_info[i_c].energy = 0.5 * y**2 * sap_info[i_c].R
+
+
+@ti.data_oriented
+class RigidConstraint(BaseConstraint):
+    """
+    Rigid body constraints in SAPCoupler. Currently only support joint equality constraints.
+    """
+
+    def __init__(
+        self,
+        simulator: "Simulator",
+        stiffness: float = 1e8,
+        beta: float = 0.1,
+    ) -> None:
+        super().__init__(simulator, stiffness, beta)
+        self.rigid_solver = simulator.rigid_solver
+        self.constraint_solver = simulator.rigid_solver.constraint_solver
+        self.max_constraints = simulator.rigid_solver.n_equalities * self._B
+        self.n_constraints = ti.field(gs.ti_int, shape=())
+        self.constraint_type = ti.types.struct(
+            batch_idx=gs.ti_int,  # batch index
+            i_dof1=gs.ti_int,  # index of the first DOF in the constraint
+            i_dof2=gs.ti_int,  # index of the second DOF in the constraint
+            sap_info=self.sap_constraint_info_type,  # SAP info for the constraint
+        )
+        self.constraints = self.constraint_type.field(shape=(self.max_constraints,))
+        self.Jt = ti.field(gs.ti_float, shape=(self.max_constraints, self.rigid_solver.n_dofs))
+        self.M_inv_Jt = ti.field(gs.ti_float, shape=(self.max_constraints, self.rigid_solver.n_dofs))
+        self.W = ti.field(gs.ti_float, shape=(self.max_constraints,))
+
+    @ti.kernel
+    def build_constraints(
+        self,
+        equalities_info: array_class.EqualitiesInfo,
+        joints_info: array_class.JointsInfo,
+        static_rigid_sim_config: ti.template(),
+    ):
+        self.n_constraints[None] = 0
+        self.Jt.fill(0.0)
+        # TODO: Maybe support different constraints for each batch in the future. For example
```

Comment:
What is this "For example" at the end of the sentence? You start next line with an uppercase.